Add target_toolchain_roots and make sysroot more hermetic

AustinSchuh opened the pull request:
It is useful when cross compiling to have 1 LLVM build used as the compiler, and multiple others used for their libc++/target libraries.

This concept forces us to be much more robust about what comes from the target toolchain, and what comes from the compiler.  Leave toolchain_roots for the compiler roots, but add a new flag, target_toolchain_roots with the same semantics for target toolchains. This necesitates being much more clear internally about what comes from what.

This isn't fully enough to make everything hermetic, clang still has some autodetection logic left over when using sysroots.  We want to very explicity specify the resource directory it should use, as well as where the standard C/C++ libraries come from so it finds only what we want it to find.

When pulling from a sysroot, we need to know the version of stdc++ to find includes.  Use the stdlib variable for that by supporting "stdc++-12" for example, and parsing the number out.

AustinSchuh committed:
Add target_toolchain_roots and make sysroot more hermetic

It is useful when cross compiling to have 1 LLVM build used as the
compiler, and multiple others used for their libc++/target libraries.

This concept forces us to be much more robust about what comes from the
target toolchain, and what comes from the compiler.  Leave
toolchain_roots for the compiler roots, but add a new flag,
target_toolchain_roots with the same semantics for target toolchains.
This necesitates being much more clear internally about what comes from
what.

This isn't fully enough to make everything hermetic, clang still has
some autodetection logic left over when using sysroots.  We want to very
explicity specify the resource directory it should use, as well as where
the standard C/C++ libraries come from so it finds only what we want it
to find.

When pulling from a sysroot, we need to know the version of stdc++ to
find includes.  Use the stdlib variable for that by supporting
"stdc++-12" for example, and parsing the number out.

AustinSchuh commented:
https://github.com/bazelbuild/rules_cc/pull/304 needs to land and be available before this can merge.  But I would appreciate feedback on this before then as that could take a bit.

We used features before for asan/msan/cuda/etc, but I think we can use flags + select instead and avoid needing to change rules_cc again.  Feedback welcome on the best way to do all that.

AustinSchuh committed:
Merge branch 'master' into target_toolchain_roots

AustinSchuh committed:
Switch to using fancier isystem flags

AustinSchuh committed:
Merge remote-tracking branch 'origin/master' into target_toolchain_roots

# Conflicts:
#	toolchain/cc_toolchain_config.bzl

AustinSchuh committed:
Fix lint issues

AustinSchuh committed:
Fix KeyError in llvm_toolchain macro

AustinSchuh committed:
Make target_toolchain_roots and toolchain_roots optional attributes

AustinSchuh commented:
Thanks @dzbarsky.  Your goals sound similar, but I've got numerous yocto sysroots I need to cross compile for, and will need to add CUDA support for those sysroots too.  The docs on your repo suggest that is not the direction you want to go.

Long long ago (2019 or earlier), we patched the old grailbio toolchains to do all this, and never got around to upstreaming our changes.  This is one patch in a series attempting to get us back inline with upstream.

AustinSchuh committed:
Use @@ for Bzlmod toolchain roots to avoid invalid Label errors

AustinSchuh committed:
Move label:

AustinSchuh committed:
Remove use_lld reference from cc_toolchain_config.bzl

AustinSchuh committed:
Add missing multiarch element to target os/arch tuples in cc_toolchain_config.bzl

AustinSchuh committed:
Further fixes

AustinSchuh committed:
Fix isystem-after

AustinSchuh committed:
Support absolute paths  again

AustinSchuh committed:
Fix libc++ cross compilation

AustinSchuh committed:
idirafter

Signed-off-by: Austin Schuh <austin.linux@gmail.com>

AustinSchuh committed:
Merge remote-tracking branch 'origin/master' into target_toolchain_roots

AustinSchuh committed:
Merge remote-tracking branch 'origin/master' into target_toolchain_roots

AustinSchuh committed:
Merge branch 'master' into target_toolchain_roots

helly25 commented on toolchain/internal/configure.bzl:
```diff
@@ -159,10 +165,35 @@ def llvm_config_impl(rctx):
-    system_llvm = False
-    if _is_absolute_path(toolchain_root):
+    system_llvm = _is_absolute_path(toolchain_root)
+    if system_llvm:
```
why not:
use_absolute_paths_llvm = _is_absolute_path(toolchain_root)

helly25 commented on toolchain/internal/configure.bzl:
```diff
@@ -740,6 +796,8 @@ def _is_remote(rctx, exec_os, exec_arch):
 
 def _convenience_targets_str(rctx, use_absolute_paths, llvm_dist_rel_path, llvm_dist_label_prefix, exec_dl_ext):
+    # TODO: This doesn't really deal with cross compilation very well.  There can be multiple
+    # targets for 1 toolchain, and the target platform matters, not the exec platform.
```
So in most cases there should be a fail here or what do I get from this comment?

AustinSchuh replied:
I cleaned up the docs to be more clear.  This was trying to say that these are convenience targets, which make sense for builds on the target, but not cross compilation.  If someone can articulate what they should look like for cross compilation, we can fix them.

helly25 commented on toolchain/rules.bzl:
```diff
@@ -53,6 +53,9 @@ def llvm_toolchain(name, **kwargs):
         llvm(name = name + "_llvm", **llvm_args)
 
+    if not kwargs.get("target_toolchain_roots"):
+        kwargs["target_toolchain_roots"] = kwargs.get("toolchain_roots", {})
```
Can we not do this, so we know whether it is explicitly set or unset?

AustinSchuh replied:
Make sense, done.  I was trying to make the default case match the target_toolchain_root case, but I found a cleaner way downstream.

helly25 commented on toolchain/cc_toolchain_config.bzl:
```diff
@@ -38,13 +38,14 @@ def cc_toolchain_config(
         target_os,
         target_system_name,
         toolchain_path_prefix,
+        target_toolchain_path_prefix,
```
We require 'target_toolchain_path_prefix' to end in '/', just like other path prefixes. But it occurred to me thta we never actually verify that, do we?

AustinSchuh replied:
Verification added!  But yes, we do.  I don't think those were validated either.

AustinSchuh committed:
Round 1 of comments, rerunning tests

AustinSchuh committed:
Revert isystem change

AustinSchuh committed:
Fix #include_next breakage in rules_foreign_cc on Bazel 8

When we pass -cxx-isystem with a relative toolchain path, clang's auto-detection of libc++ adjacent to the clang binary still adds the same directory with an absolute path (computed from getcwd()+argv[0]). Clang's path-string dedup treats these as different entries, so the toolchain's c++/v1 appears twice in the include search list.

This breaks #include_next in libc++ wrappers like <stdlib.h>: the 'next' header is the same wrapper at the absolute-path duplicate (skipped by its own header guard), so the C library's stdlib.h is never reached, leaving ldiv_t, size_t, etc. undeclared.

Adding -nostdinc++ disables clang's auto-detection so only our explicit -cxx-isystem path is in the search list, restoring correct #include_next semantics for both same-toolchain and cross-compile cases. Verified locally on macOS with Bazel 8.7.0 against //foreign:pcre, which previously failed with 'unknown type name ldiv_t' and now builds cleanly.

Signed-off-by: Austin Schuh <austin.linux@gmail.com>

AustinSchuh committed:
Merge branch 'master' into target_toolchain_roots

helly25 commented:
> OK, looks like everything passes again now. More changes keep getting added to main as this review stays outstanding.
> 
> I'd like to add some tests for full on cross compilation as well, but those need a couple more changes on top of this. I figure this change is big enough as is, though I'm happy to make it whatever size makes your lives easier.

Finally merged :-)

Yeah, we pretty much need those tests to guarantee things stay in working order. I assume a small number of tests should do it that involve Linux, Windows and MacOS. But I also expected that you were not fully ready for those -:)
